tests/test_autosave.py

```diff
@@ -0,0 +1,418 @@
+            saved = yaml.full_load(f)
+        assert "SAVEVAL" in saved
+        assert "SAVEVAL.DISA" in saved
+        assert "DONTSAVEVAL" not in saved
+        assert "DONTSAVEVAL.SCAN" in saved
+
+
+def test_stop_event(tmp_path):
+    autosave.configure(tmp_path, DEVICE_NAME)
+    builder.aOut("DUMMYRECORD", autosave=True)
+    worker = threading.Thread(
+        target=autosave.Autosave._loop,
+    )
+    try:
+        worker.daemon = True
+        worker.start()
+        assert not autosave.Autosave._stop_event.is_set()
+        assert worker.is_alive()
+        autosave.Autosave._stop()
+        assert autosave.Autosave._stop_event.is_set()
+    finally:
+        worker.join(timeout=1)
+
+
+@pytest.mark.parametrize(
+    "timestamped,regex",
+    [
+        (False, r"^" + DEVICE_NAME + r"\.softsav_[0-9]{6}-[0-9]{6}$"),
+        (True, r"^" + DEVICE_NAME + r"\.softsav\.bu$"),
+    ],
+)
+def test_backup_on_load(existing_autosave_dir, timestamped, regex):
+    autosave.configure(
+        existing_autosave_dir, DEVICE_NAME, timestamped_backups=timestamped
+    )
+    # backup only performed if there are any pvs to save
+    builder.aOut("SAVED-AO", autosave=True)
+    autosave.load_autosave()
+    backup_files = list(existing_autosave_dir.glob("*.softsav_*"))
+    # assert backup is <name>.softsav_yymmdd-HHMMSS or <name>.softsav.bu
+    any(re.match(regex, file.name) for file in backup_files)
+    if not timestamped:
+        # test that existing .bu file gets overwritten
+        with open(existing_autosave_dir / f"{DEVICE_NAME}.softsav.bu") as f:
+            state = yaml.full_load(f)
+            assert "OUT-OF-DATE-KEY" not in state
+            assert "SAVED-AO" in state
+
+
+def test_autosave_key_names(tmp_path):
+    builder.aOut("DEFAULTNAME", autosave=True)
+    builder.aOut("DEFAULTNAMEAFTERPREFIXSET", autosave=True)
+    autosave.configure(tmp_path, DEVICE_NAME)
+    autosaver = autosave.Autosave()
+    autosaver._save()
+    with open(tmp_path / f"{DEVICE_NAME}.softsav", "r") as f:
+        saved = yaml.full_load(f)
+    assert "DEFAULTNAME" in saved
+    assert "DEFAULTNAMEAFTERPREFIXSET" in saved
+
+
+def test_context_manager(tmp_path):
+    builder.aOut("MANUAL", autosave=True, autosave_fields=["EGU"])
+    with autosave.Autosave(True, ["PINI"]):
+        builder.aOut("AUTOMATIC")
+        builder.aOut(
+            "AUTOMATIC-OVERRIDDEN", autosave=False, autosave_fields=["SCAN"]
+        )
+    autosave.configure(tmp_path, DEVICE_NAME)
+    with patch(
+        "softioc.device.ProcessDeviceSupportCore.get_field", return_value="0"
+    ):
+        autosaver = autosave.Autosave()
+        autosaver._save()
+        with open(tmp_path / f"{DEVICE_NAME}.softsav", "r") as f:
+            saved = yaml.full_load(f)
+        assert "MANUAL" in saved
+        assert "MANUAL.EGU" in saved
+        assert "AUTOMATIC" in saved
+        assert "AUTOMATIC.PINI" in saved
+        assert "AUTOMATIC-OVERRIDDEN" in saved
```

Comment:
Is it expected that the Context Manager's `True` would override the `autosave=False`? Or is there some other reason this item is in the saved list?

Reply:
That's how it works currently yes, I suppose we could find a way to use some True/False/None logic if we wanted an explicit calling of autosave=False to override the context manager?

Reply:
I see. My view would be that setting `autosave=False` on the record creation function should override everything else - it's the principle of least surprise to take its value over any other.

However, I can see that may be difficult to actually achieve. I can also see that it seems very unlikely anyone will ever actually try and do that. If there's an easy way to at least print/raise an error in this case then do that, otherwise it is a rather confusing behaviour. @Araneidae, any thoughts?

Reply:
Can roll back if required but just pushed something to address this: we call add_pv_to_autosave on all PVs at init now, and class variables are set in the context manager which are used to override the the arguments "autosave" and "autosave_fields" if not specified, so an explicit autosave=False will get applied. It also means we don't have to use LookupRecordList any more.

But I did realise that this newer implementation, and the LookupRecordList implementation would not be thread safe, in the sense that if a separate thread is creating PVs while another thread is creating PVs inside a context manager, we may accidentally track fields that are not specified. I'm not sure if anyone would write their IOC that way, but may be something worth trying to guard against.